### DISCUSSION TITLE
Serial server and client - No reply from server.

### DISCUSSION THREAD

**Discussion by 1038642892** [pymodbus-dev/pymodbus#1121] [Help] [answered]
### Scenario
I'd like to use the server and console REPL to test serial communication locally on Linux (WSL2) with a socat provided link between two virtual serial ports.

### Scenario diagram
Server <-> /dev/pts/7 <-> socat <-> /dev/pts/6 <-> Client

### Issue

The server does not reply to the command... shown below.

### Server init command
`pymodbus.server --web-port 8881 run --modbus-server serial --framer rtu --modbus-port /dev/pts/7 --unit-id 1`

### Client init command and REPL commands
`pymodbus.console serial --method rtu --port /dev/pts/8 --baudrate 9600 --bytesize 8 --parity N --stopbits 1`
```
> client.connect
true

> client.read_coils slave=1 address=1 count=1
{
    "original_function_code": "1 (0x1)",
    "error": "[Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received)"
}
```

### Socat pipe in between
This command below creates a link between two virtual serial ports, and prints messages out in hex format.
This has been proven to work with another piece of code (i can share this if required).
`socat -x -v -d -d PTY,raw,echo=0,ispeed=b9600 PTY,raw,echo=0,ospeed=b9600`
`sudo chown user:user /dev/pts/7 /dev/pts/8`

This is the output (requesting read coil address 1, count 1, slave 1.
```
2022/10/20 10:58:53 socat[1873] N PTY is /dev/pts/7
2022/10/20 10:58:53 socat[1873] N PTY is /dev/pts/8
2022/10/20 10:58:53 socat[1873] N starting data transfer loop with FDs [5,5] and [7,7]
< 2022/10/20 11:09:53.296277  length=8 from=0 to=7
 01 01 00 01 00 01 ac 0a                          ........
--
```

You'll see there's no reply.

Any support is appreciated to try to troubleshoot why i'm not seeing a message/correct message back from server.

**Comment by janiversen** [comment]
Socat works nicely, I use it daily. You can test serial without socat too, see test/test_examples.py

@dhoomakethu is the expert on REPL, I am concentrating on the library itself and as you can see from our test setup the library as such works.

**Reply by 1038642892** [reply]
Thank you for your message, i will check out the test examples.

**Comment by janiversen** [comment]
I will however next week start using REPL client to test some a new datastore feature that allows you to build device simulations, so maybe I will see and fix your problem.

**Comment by dhoomakethu** [comment]
@1038642892 what version of pymodbus are you using ?

I just gave a try with latest on `dev` and it seems to be working
<img width="896" alt="image" src="https://user-images.githubusercontent.com/9276974/196933300-1bba5223-0028-4ef3-866c-d35a37e658d3.png">
<img width="661" alt="image" src="https://user-images.githubusercontent.com/9276974/196933372-a5dc3cdd-36f0-4cb1-9dee-d34a03a3657c.png">

**Reply by dhoomakethu** [reply]
For your reference this is the `socat` caommand I use 
```
socat -d -d PTY,link=/tmp/ptyp0,raw,echo=0,ispeed=9600 PTY,link=/tmp/ttyp0,raw,echo=0,ospeed=9600
```

**Reply by dhoomakethu** [reply]
Looks like you are missing `--timeout` option in the client , add a timeout like `--timeout 2` and see if that helps.

**Reply by 1038642892** [reply]
I think we have different socat versions
`sudo socat -d -d PTY,link=/tmp/ptyp0,raw,echo=0,ispeed=9600 PTY,link=/tmp/ttyp0,raw,echo=0,ospeed=9600`
```
2022/10/20 12:22:18 socat[1905] N PTY is /dev/pts/7
2022/10/20 12:22:18 socat[1905] E cfsetispeed(0x562559ec39c0, 9600): Invalid argument
2022/10/20 12:22:18 socat[1905] N exit(1)
```
But changing to this command (b9600) seems to create the /dev/pts/7 and 8 with symlink /tmp/ttyp0 and ptyp0 just fine.
`sudo socat -d -d PTY,link=/tmp/ptyp0,raw,echo=0,ispeed=b9600 PTY,link=/tmp/ttyp0,raw,echo=0,ospeed=b9600`
```
2022/10/20 12:22:33 socat[1907] N PTY is /dev/pts/7
2022/10/20 12:22:33 socat[1907] N PTY is /dev/pts/8
2022/10/20 12:22:33 socat[1907] N starting data transfer loop with FDs [5,5] and [7,7]
```

Here's the new server and client test with your requested change to timeout and similar params to yours.

`pymodbus.server --verbose --web-port 8881 run --modbus-server serial --framer rtu --modbus-port /tmp/ttyp0 --unit-id 1 --random 2 --baudrate 9600`
`pymodbus.console serial --method rtu --port /tmp/ptyp0 --baudrate 9600 --timeout 2`

![image](https://user-images.githubusercontent.com/102810494/196937434-6b520f7f-b5fb-4c1d-a8ad-7af235aa7a34.png)
![image](https://user-images.githubusercontent.com/102810494/196937442-2e912ff9-68e9-4ea1-b3da-a492abcef2f9.png)

For reference my socat version
```
socat -V
socat by Gerhard Rieger and contributors - see www.dest-unreach.org
socat version 1.7.3.3 on Oct 26 2019 17:42:04
   running on Linux version #1 SMP Fri Apr 2 22:23:49 UTC 2021, release 5.10.16.3-microsoft-standard-WSL2, machine x86_64
features:
  #define WITH_STDIO 1
  #define WITH_FDNUM 1
  #define WITH_FILE 1
  #define WITH_CREAT 1
  #define WITH_GOPEN 1
  #define WITH_TERMIOS 1
  #define WITH_PIPE 1
  #define WITH_UNIX 1
  #define WITH_ABSTRACT_UNIXSOCKET 1
  #define WITH_IP4 1
  #define WITH_IP6 1
  #define WITH_RAWIP 1
  #define WITH_GENERICSOCKET 1
  #define WITH_INTERFACE 1
  #define WITH_TCP 1
  #define WITH_UDP 1
  #define WITH_SCTP 1
  #define WITH_LISTEN 1
  #define WITH_SOCKS4 1
  #define WITH_SOCKS4A 1
  #define WITH_PROXY 1
  #define WITH_SYSTEM 1
  #define WITH_EXEC 1
  #undef WITH_READLINE
  #define WITH_TUN 1
  #define WITH_PTY 1
  #define WITH_OPENSSL 1
  #undef WITH_FIPS
  #define WITH_LIBWRAP 1
  #define WITH_SYCLS 1
  #define WITH_FILAN 1
  #define WITH_RETRY 1
  #define WITH_MSGLEVEL 0 /*debug*/
```

**Comment by janiversen** [comment]
Just for info, in respect of REPL dev is the same a release 3.0.0 for now.

**Reply by 1038642892** [reply]
I have also found that using non REPL - code example has the same issue - no reply from server.

In my screenshot you can see i've enabled --verbose on server, is this DEBUG message the only one expected with the --verbose or should I see further messages?
`DEBUG async_io:860 Exception while create - 'ModbusSingleRequestHandler' object is not callable`

**Comment by janiversen** [comment]
This is a bug we solved quite a while ago, seems you are using an old version.

**Reply by 1038642892** [reply]
OK I will fetch source and try with that. Thank you

**Comment by 1038642892** [comment, marked as answer]
I was using the version available in pip.

I've removed pymodbus-3.0.0 and tried from source (3.0.1.dev0).

This seems to have fixed it. Thank you for your time @janiversen and @dhoomakethu. I'm impressed with the pymodbus tooling.

See below for installation steps used:
```
virtualenv testenv
source testenv/bin/activate
git clone <URL>
pip install -e .
```

**Comment by janiversen** [comment]
thanks for the report, this gives me a good reason to speedup 3.0.1
